### PR TITLE
[JIT] Add recursive scripting for class type module attributes

### PR DESCRIPTION
**Pull request by SplitInfinity**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45234 [JIT] Add recursive scripting for class type module attributes**

**Summary**
This commit modifies type inference (used by the module scripting code)
so that it tries to script the type of any class instances that it
encounters. This enables recursive, automatic scripting of class type
module attributes.

**Test Plan**
This commit adds a test case for this to `TestClassType`.

Differential Revision: [D23971883](https://our.internmc.facebook.com/intern/diff/D23971883)

### CODE REVIEW

File: torch/csrc/jit/python/pybind_utils.h
Comment (eellison) [inline torch/csrc/jit/python/pybind_utils.h:310]: i don't think we actually resolve module types to annotations , but the !is_module cant hurt i suppose

**Review by eellison — APPROVED** [review]
Nice

File: torch/csrc/jit/python/pybind_utils.h
Comment (SplitInfinity) [reply torch/csrc/jit/python/pybind_utils.h:310]: `TestAsync.test_async_script_trace` breaks without this because the submodule in question extends `ScriptModule`, not `nn.Module`, so I think that causes `resolveTypeFromObject` to return the module's type.

File: torch/_jit_internal.py
Comment (gmagogsfm) [inline torch/_jit_internal.py:226]: Could you clarify a bit on why this is necessary? Thanks

Comment (SplitInfinity) [reply torch/_jit_internal.py:226]: In `TestQuantizeFx.test_qat_and_script`, we try to script a `GraphModule` which has a `torch.fx.Graph` as an attribute. `Graph` is a regular object, not a `Module`, but we shouldn't try to script it.

I wonder if using the API I introduced in #45262 is a better solution here. We could ignore the attribute by adding `__ignored_attributes__` to `GraphModule` rather than doing this. What do you think?

File: torch/jit/_recursive.py
Comment (SplitInfinity) [inline torch/jit/_recursive.py:23]: @eellison As you predicted there are some attributes now getting scripted that shouldn't be. I would prefer to ignore them by modifying `__jit_ignored_attributes__` whereever they are added but I can't find the code that adds the `qconfig` attribute. I think `_qconfig_map` is added here: https://github.com/pytorch/pytorch/blob/master/torch/quantization/fx/quantize.py#L429

**Comment by eellison** [comment]
Is there anyway we could try to cherry-pick these two commits into 1.7? There will be users who have to ignore properties & ideally we don't have API churn for how to do that

**Comment by SplitInfinity** [comment]
> Is there anyway we could try to cherry-pick these two commits into 1.7? There will be users who have to ignore properties & ideally we don't have API churn for how to do that

I would like to do that, but I am waiting on a review for #45262 before I can land this and that PR and have them cherry picked.

**Comment by eellison** [comment]
@SplitInfinity accepted, sorry re-review by me slipped

**Comment by eellison** [comment]
Too late to get this into 1.7 ? I was just trying to use class types, and got bit by this

**Comment by SplitInfinity** [comment]
I'm still dealing with internal failures: D23971883 :'(
It's a slow and painful process

**Comment by eellison** [comment]
Gotcha, makes sense.

**Comment by SplitInfinity** [comment]
I'm re-requesting review on this because I made significant changes after the first version caused a lot of internal failures that were hard to track down and fix using `__jit_ignored_attributes__`. This first version tried to script all module attributes that are instances of class types (except for those present in `__jit_ignored_attributes__` and some standard exempted types like `Exception`), and this just caused way too many internal tests to fail. It even ended up trying to recursively script a class defined in a library that's external to fbcode.

In response, I modified the approach so that if the attempt to script the class of a potential class type attribute fails, we add that attribute as a failed attribute to the module's concrete type (which is what happens today with attributes whose classes are not annotated with `@torch.jit.script`), even if that attribute is *not* present in `__jit_ignored_attributes__`. If the attempt to script the class succeeds, we add the attribute to the module's concrete type with the JIT type produced by scripting its class.

This has drastically reduced the number of internal failures and should be easier to deal with now.

File: torch/csrc/distributed/rpc/py_rref.cpp
Comment (SplitInfinity) [inline torch/csrc/distributed/rpc/py_rref.cpp:90]: These changes are necessary because `tryToInferType` now automatically scripts classes that can be scripted, which can result in scripting classes that the user did not intend to be scripted. In a JIT context, this makes more sense because you'd probably want your module attributes to all be scripted. However, at present, you can construct an `RRef` with a regular Python class instance too, and because this code uses `jit::tryToInferType`, due to the other changes in this PR (automatic scripting of class types), it is no longer possible to do this.

This small patch checks if the type of the object you are creating an `RRef` of could be a compiled as a class type and if it can, it checks if it already has been compiled or not. If it has, that means the class has been decorated with `@torch.jit.script` and the user intended it to be a TorchScript class. If not, it means that the user did not intend to script it, so we short-circuit and return `PyObjectType::get()`.

Comment (wanchaol) [reply torch/csrc/distributed/rpc/py_rref.cpp:90]: curious if we should just enable recursive scripting to automatically compile the class? I think `tryToInferType` shouldn't automatically compile the class if user don't provide the `torch.jit.script` annotation, which make more sense to me that only explicit annotation marking the user intention to compile the class and use the object in their code. Do we have a use case why `tryToInferType` has to compile the class?

Comment (SplitInfinity) [reply torch/csrc/distributed/rpc/py_rref.cpp:90]: The motivation for the PR is for automatically scripting module attributes that are classes. We do recursive scripting automatically for submodules, so this is an extension of that. I think that makes sense, and I think the problem is that this `RRef` code is using a type inference function that is meant for JIT purposes.

**Comment by eellison** [comment]
@SplitInfinity I think the behavior is reasonable enough, the thing that I'm concerned abt is that by trying to compile a class and failing, we don't know if we've left any parts of the system in a bad intermediate state.

File: torch/csrc/distributed/rpc/py_rref.cpp
Comment (wanchaol) [reply torch/csrc/distributed/rpc/py_rref.cpp:90]: Yeah I agree that we would like to apply this to module attributes that are classes. But I thought we already automatically compile it when the class attribute get annotated ([here](https://github.com/pytorch/pytorch/blob/75bf5f2b59148ea8b6a3844248bcc645dd80e0c8/torch/jit/annotations.py#L332))? it seems what this part of code is doing is to try to recursively compile the class that does not have class annotations?

Comment (SplitInfinity) [reply torch/csrc/distributed/rpc/py_rref.cpp:90]: Yes, it is trying to recursively compile the class of a module attribute even if it is not annotated.

**Comment by SplitInfinity** [comment]
That's a good point. The only internal state that I'm aware of is `_script_classes` in `_state.py`, a dictionary to which we add a class after it has been successfully compiled.
```
def _compile_and_register_class(obj, rcb, qualified_name):
    ast = get_jit_class_def(obj, obj.__name__)
    defaults = torch.jit.frontend.get_default_args_for_class(obj)
    torch._C._jit_script_class_compile(qualified_name, ast, defaults, rcb)
    torch.jit._state._add_script_class(obj, qualified_name)
```

I think we won't add a class that fails to script to the registry, because `_jit_script_class_compile` will throw.

**Comment by eellison** [comment]
@SplitInfinity that's the python caching, but I wouldn't be surprised if theres all sorts of C++ state that might get set in a bad intermediary state. 

I'd be curious how many existing script codebases would break with adding recursive scripting of classes on by default. Like torchvision for example.

Throwing out crazy suggestions here - would it be possible to upload a version of this diff internally where we try to script the class, and if it fails log the qualified name of the class and attribute somewhere ? Then maybe we could batch apply all the ignores.

**Comment by eellison** [comment]
@suo @zdevito anyone want to  weigh in specifically on catching exceptions  when trying to script a class, and if that poses any problems ?

**Comment by SplitInfinity** [comment]
@eellison 
I noticed today that we do something similar for function attributes:
```
        # Handle Python function attributes
        if inspect.isfunction(value):
            try:
                scripted_fn = torch.jit.script(value)
                concrete_type_builder.add_function_attribute(
                    name,
                    torch._C._jit_try_infer_type(scripted_fn),
                    value)
            except Exception as e:
                # If we fail to script the function, it isn't a hard error.
                # Instead, we will add it to the list of attributes we failed
                # to convert, with the compilation error.
                hint = ("(This function exists as an attribute on the Python module, "
                        "but we failed to compile it to a TorchScript function. "
                        "\nThe error stack is reproduced here:\n{}").format(e)
                concrete_type_builder.add_failed_attribute(name, hint)
                pass
```
Not exactly the same as scripting a class though.

**Comment by eellison** [comment]
Cool, yea i think we should go ahead and land this in some form then.

**Review by eellison — APPROVED** [review]
Looks good to me - would love to get additional input from @gmagogsfm or @suo but I think we should go forward with this, a lot of users are confused by current behavior and this is a nice usability improvement

**Pull request closed without merging**
